### PR TITLE
refactor!: require systems to be `complete`d before creating `XProblem`/`XProblemExpr`

### PR DESCRIPTION
**Pull request by AayushSabharwal**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

**Comment by AayushSabharwal** [comment]
CI won't pass here because of the states->unknowns change, but it runs locally

**Comment by AayushSabharwal** [comment]
As before, CI passes except not all of `test/optimizationsystem.jl` runs because the changes required to make this work are split across two PRs

**Comment by ChrisRackauckas** [comment]
Resolve conflicts

File: docs/src/basics/Events.md [hunk 1]
```diff
@@ -73,7 +73,7 @@ function UnitMassWithFriction(k; name)
     ODESystem(eqs, t; continuous_events = [v ~ 0], name) # when v = 0 there is a discontinuity
 end
 @named m = UnitMassWithFriction(0.7)
```
Comment (ChrisRackauckas) [inline docs/src/basics/Events.md:75]: Can we update all of these to `@mtkbuild` instead?

**Pull request merged**